bigint/src/lib.rs

```diff
@@ -1349,6 +1356,47 @@ impl Integer for BigUint {
     }
 }
 
+fn high_bits_to_u64(v: &BigUint) -> u64 {
+    match v.data.len() {
+        0   => 0,
+        1   => v.data[0] as u64,
+        _   => {
+            let mut bits = v.bits();
+            let mut ret = 0u64;
+            let mut ret_bits = 0;
+
+            for d in v.data.iter().rev() {
+                let digit_bits = (bits + big_digit::BITS - 1) %
+                    big_digit::BITS + 1;
```

Comment:
Why are you adding `BITS` first?  That should be a modular no-op.  Only reason I can think is if `bits` were 0, to avoid subtraction underflow, but that should never be the case AFAICT.